src/server/saved_objects/migrations/core/build_active_mappings.ts

```diff
@@ -38,14 +48,66 @@ export function buildActiveMappings({
   properties: MappingProperties;
 }): IndexMapping {
   const mapping = defaultMapping();
+
+  properties = validateAndMerge(mapping.doc.properties, properties);
+
   return _.cloneDeep({
     doc: {
       ...mapping.doc,
-      properties: validateAndMerge(mapping.doc.properties, properties),
+      properties,
+      _meta: {
+        migrationMappingHash: md5Values(properties),
+      },
     },
   });
 }
 
+/**
+ * Diffs the actual vs expected mappings. The properties are compared using md5 hashes stored in _meta, because
+ * actual and expected mappings *can* differ, but if the md5 hashes stored in actual.doc._meta.migrationMappingHash
+ * match our expectations, we don't require a migration. This allows ES to tack on additional mappings that Kibana
+ * doesn't know about or expect, without triggering continual migrations.
+ */
+export function diffMappings(
+  actual: IndexMapping,
+  expected: IndexMapping
+): MappingDiff | undefined {
+  if (actual.doc.dynamic !== expected.doc.dynamic) {
+    return { changedProp: 'doc.dynamic' };
+  }
+
+  if (!actual.doc._meta || !actual.doc._meta.migrationMappingHash) {
+    return { changedProp: 'doc._meta' };
+  }
+
+  const changedProp = findChangedProp(
+    actual.doc._meta.migrationMappingHash,
+    expected.doc._meta!.migrationMappingHash
+  );
+
+  return changedProp ? { changedProp: `doc.properties.${changedProp}` } : undefined;
+}
+
+// Convert an object to an md5 hash string, using rison so we get a stable serialization
+function md5Object(obj: any) {
+  return crypto
+    .createHash('md5')
+    .update(rison.encode(obj))
```

Comment:
Hm. That's a good point. I didn't feel like thinking through the edge cases of a canonical JSON implementation, but all we probably need is sorted keys. I'm a fan of reducing dependencies in general.

Reply:
OK. This works. It doesn't matter whether or not it produces valid JSON, only that it produces consistent stringification of an object for hashing purposes:

```js
function canonicalStringify(obj) {
  if (Array.isArray(obj)) {
    return `[${obj.map(canonicalStringify)}]`;
  }

  if (!obj || typeof obj !== 'object') {
    return JSON.stringify(obj);
  }

  const keys = Object.keys(obj);

  if (!keys.length) {
    return JSON.stringify(obj);
  }

  const sortedObj = keys
    .sort((a, b) => a.localeCompare(b))
    .map(k => `${k}: ${canonicalStringify(obj[k])}`);

  return `{${sortedObj}}`;
}
```

My vote is rison, but my second vote is something like ^^^ instead of adding more deps to package.json.

Thoughts?

Reply:
There's no reason for using zipObject, just sort the entries and JSON stringify  the entries as an array?

Reply:
Actually, what you have there looks like it works. Want to through it in its own file and put some simple unit tests in for it?

Reply:
Oh, you tested it through the diffMappings api. Nevermind

Reply:
Oh, you tested it through the diffMappings api. Nevermind